### DISCUSSION TITLE
Which target was selected in the Null / always transient transition ?

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2114] [Q&A] [answered]
Hello 👋 

One of my states validates a file and chooses to go back to an `idle` state, when validation fails or perform the file upload if the validation succeeds. My state looks like this

```js
export const ValidateState = {
   meta: {
      message: "Validation State performs file validations pre-upload",
   },
   type: "atomic",
   always: [
      {
         target: "upload",
         cond: "isValidFileObj",
         actions: ["updateCurrentFile"],
      },
      { target: "diagnostics", actions: ["cleanup"] },
   ],
   exit: ["logValidationResult"],
};
```
When inspecting `event` param of The exit action `logValidationResult` , the type is `""`. 
Is there a way to get which **target** was chosen ?

**Comment by Andarist** [comment, marked as answer]
> When inspecting event param of The exit action logValidationResult , the type is "".

This will be fixed in the next major release - the event object will be "preserved". Whatever event has triggered the original transitions should be preserved, regardless of how many `always` hops are taken in response to it.

> Is there a way to get which target was chosen ?

No - it seems like a separate question though.

**Reply by awreccan** [reply]
> Whatever event has triggered the original transitions should be preserved

@Andarist has there been any work done on this so far? I'd like to consume this feature even if it's at PR stage.

**Reply by davidkpiano** [reply]
> @Andarist has there been any work done on this so far? I'd like to consume this feature even if it's at PR stage.

I believe so - check out the `next` branch.
